Earthfile

```diff
@@ -39,241 +64,325 @@ setup:
-  LABEL org.label-schema.vcs-url="https://github.com/space-ros/docker-images"
+  LABEL org.label-schema.vcs-url="https://github.com/space-ros/space-ros"
   LABEL org.label-schema.vcs-ref=${VCS_REF}
 
-  COPY +build/install ${SPACEROS_DIR}/install
-  COPY +spaceros-artifacts/exact.repos ${SPACEROS_DIR}/exact.repos
-  RUN rm -r src
+  # Add missing dependencies
+  RUN apt install -y libspdlog-dev \
+      python3-numpy \
+      python3-rosdistro \
+      tzdata \
+      sudo \
+      python3-colcon-common-extensions \
+      python3-rosdep \
+    && pip3 install pyyaml \
+      lark \
+      packaging \
+      netifaces \
+      catkin_pkg \
+      psutil
+
+  # Prepare the image
+  RUN mkdir -p ${SPACEROS_DIR}
+  COPY +rosdep/rosdeps.sh ${SPACEROS_DIR}/rosdeps.sh
+  RUN bash ${SPACEROS_DIR}/rosdeps.sh
+
+###############################################################################
+### Prepare Image Stage
+# This stage is responsible for preparing the image with the ROS 2 workspace.
+# The image is saved only if the build-test stage is successful.
+###############################################################################
+image:
+  ARG --required IMAGE_VARIANT
+  FROM +prepare-image
+
+  COPY +build/spaceros ${SPACEROS_DIR}
+  COPY +sources/exact.repos ${SPACEROS_DIR}/spaceros.repos
+
+  # Post Installation
+  DO +POST_INSTALLATION --IMAGE_VARIANT=${IMAGE_VARIANT}
 
-  WORKDIR ${HOME_DIR}
-  COPY docker/entrypoint.sh /ros_entrypoint.sh
-  ENTRYPOINT ["/ros_entrypoint.sh"]
-  CMD ["bash"]
-  SAVE IMAGE --push osrf/space-ros:latest osrf/space-ros:$tag
+  # Clear Apt and Pip cache
+  RUN rm -rf /var/lib/apt/lists/* /tmp/pip-reqs /var/cache/apt/archives \
+    && apt-get clean \
+    && pip cache purge
+
+  # Add user and group
+  RUN useradd -m -s /bin/bash ${USERNAME} \
+    && chown -R ${USERNAME}:${USERNAME} ${SPACEROS_DIR} \
+    && echo "${USERNAME} ALL=(ALL) NOPASSWD:ALL" > /etc/sudoers.d/${USERNAME}
+  RUN echo "source ${SPACEROS_DIR}/setup.bash" >> ${HOME}/.bashrc
+
+  RUN chown -R ${USERNAME}:${USERNAME} ${HOME}
+
+  USER ${USERNAME}
+  WORKDIR ${HOME}
+
+  # Only save image if build-test is successful
+  IF [ "${SKIP_BUILD_TEST}" = "true" ]
+    SAVE IMAGE ${IMAGE_NAME}:${IMAGE_VARIANT}
+  ELSE
+    BUILD +build-test
+    SAVE IMAGE ${IMAGE_NAME}:${IMAGE_VARIANT}
+  END
+
+###############################################################################
+### Post Installation Stage
+# This stage is responsible for cleaning up the workspace and installing
+# additional dependencies based on the image variant.
+###############################################################################
+POST_INSTALLATION:
+  FUNCTION
+  ARG --required IMAGE_VARIANT
+
+  # Remove workspace if image variant is core or dev
+  IF [ "${IMAGE_VARIANT}" = "core" ] || [ "${IMAGE_VARIANT}" = "dev" ]
+    RUN rm -rf ${WORKSPACE_DIR}
+  END
+
+  # If Dev, install IKOS and excluded dependencies
+  IF [ "${IMAGE_VARIANT}" = "dev" ]
```

Comment:
I think we should probably include the exact src in the dev image as well, just so that the ikos stuff still works.

That said, eventually we'd like to move the vcs import work to outside of the container and copy the source in to avoid doing clones during the build. So maybe for now the simplest thing is to save it and copy it back in here? Will blow up the dev image size, but I think that's ok.

Reply:
I am summarizing my idea in PR directly before moving with the change.